File: exercism/python/scale-generator/scale_generator.py
class Scale:
    CHROMATIC_SHARPS = ['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#']
    CHROMATIC_FLATS = ['A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab']

    DIATONIC_SHARPS = ('A', 'a', 'C', 'G', 'f#')

    INTERVAL_DISTANCES = ('m', 'M', 'A')


    def __init__(self, tonic):
        self.tonic = tonic
        self.scale = self._get_scale_by_tonic()
        self.index = self.scale.index(self.tonic.capitalize())

    def chromatic(self):
        return self.scale[self.index:] + self.scale[:self.index]

    def interval(self, intervals):
        chromatic = self.chromatic()

        i = 0
        scale = []
        for interval in intervals:
            scale.append(chromatic[i])
            i += self.INTERVAL_DISTANCES.index(interval)+1

        return scale

    def _get_scale_by_tonic(self):
        if self.tonic in self.DIATONIC_SHARPS:
            return self.CHROMATIC_SHARPS
        else:
            return self.CHROMATIC_FLATS

File: exercism/go/scale-generator/scale_generator.go
package scale

import "strings"

var (
	sharps          = []string{"A", "B", "C", "D", "E", "G", "F#", "f#", "a", "e", "b", "c#", "g#", "d#"}
	intervals       = " mMA"
	chromaticSharps = []string{"C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"}
	chromaticFlats  = []string{"F", "Gb", "G", "Ab", "A", "Bb", "B", "C", "Db", "D", "Eb", "E"}
)

// Scale returns the diatonic scale for a given tonic and interval
func Scale(tonic, interval string) []string {
	chromatic := chromaticForTonic(tonic)

	if interval == "" {
		return chromatic
	}

	diatonic := make([]string, len(interval))

	i := startPosition(chromatic, strings.Title(tonic))

	lenChromatic := len(chromatic)
	for j, space := range interval {
		diatonic[j] = chromatic[i%lenChromatic]
		i += strings.IndexRune(intervals, space)
	}
	return diatonic
}

func chromaticForTonic(tonic string) []string {
	if contains(sharps, tonic) {
		return chromaticSharps
	}
	return chromaticFlats
}

func startPosition(scale []string, tonic string) int {
	for i, val := range scale {
		if val == tonic {
			return i
		}
	}
	return -1
}

func contains(list []string, item string) bool {
	for _, el := range list {
		if el == item {
			return true
		}
	}
	return false
}
